pkg/updates/file_test.go

```diff
@@ -0,0 +1,325 @@
+package updates
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/authzed/spicedb-operator/pkg/apis/authzed/v1alpha1"
+)
+
+func TestChannelForDatastore(t *testing.T) {
+	graph := UpdateGraph{Channels: []Channel{
+		{
+			Name:     "postgres",
+			Metadata: map[string]string{"datastore": "postgres"},
+			Nodes:    []State{{ID: "v1.0.0"}},
+		},
+		{
+			Name:     "cockroachdb",
+			Metadata: map[string]string{"datastore": "cockroachdb"},
+			Nodes:    []State{{ID: "v1.0.0"}},
+		},
+	}}
+
+	t.Run("common case", func(t *testing.T) {
+		channel, err := graph.ChannelForDatastore("cockroachdb")
+		require.Nil(t, err)
+		require.Equal(t, "cockroachdb", channel)
+		channel, err = graph.ChannelForDatastore("postgres")
+		require.Nil(t, err)
+		require.Equal(t, "postgres", channel)
+	})
+
+	t.Run("case insensitive", func(t *testing.T) {
+		channel, err := graph.ChannelForDatastore("POSTGRES")
+		require.Nil(t, err)
+		require.Equal(t, "postgres", channel)
+	})
+}
+
+func TestAvailableVersions(t *testing.T) {
+	table := []struct {
+		name          string
+		graph         *UpdateGraph
+		engine        string
+		version       v1alpha1.SpiceDBVersion
+		expectedNames []string
+		expectedErr   string
+	}{
+		{
+			name:          "empty graph",
+			graph:         &UpdateGraph{},
+			engine:        "postgres",
+			version:       v1alpha1.SpiceDBVersion{Name: "v1.0.0", Channel: "postgres"},
+			expectedNames: nil,
+			expectedErr:   `no source found for channel "postgres"`,
+		},
+		{
+			name: "graph without channel",
+			graph: &UpdateGraph{Channels: []Channel{{
+				Name:     "cockroachdb",
+				Metadata: map[string]string{"datastore": "cockroachdb"},
+				Nodes:    []State{{ID: "v1.0.0"}},
+			}}},
+			engine:        "postgres",
+			version:       v1alpha1.SpiceDBVersion{Name: "v1.0.0", Channel: "postgres"},
+			expectedNames: nil,
+			expectedErr:   `no source found for channel "postgres"`,
+		},
+		{
+			name: "graph without edges",
+			graph: &UpdateGraph{Channels: []Channel{{
+				Name:     "cockroachdb",
+				Metadata: map[string]string{"datastore": "cockroachdb"},
+				Nodes:    []State{{ID: "v1.0.1"}, {ID: "v1.0.0"}},
+			}}},
+			engine:        "cockroachdb",
+			version:       v1alpha1.SpiceDBVersion{Name: "v1.0.0", Channel: "cockroachdb"},
+			expectedNames: nil,
+			expectedErr:   "no edges or no nodes",
```

Comment:
I wonder if we can make this error more descriptive

Reply:
The error is more descriptive, I'm only checking that this string is a substring of the error

Reply:
I meant more to only say "edges" or "nodes", if missing, rather than both